orderer/common/cluster/commauth.go

```diff
@@ -0,0 +1,331 @@
+	stepRequest, cerr := BuildStepRequest(request)
+	if cerr != nil {
+		return cerr
+	}
+	return cs.StepClient.Send(stepRequest)
+}
+
+func (cs *NodeClientStream) Recv() (*orderer.StepResponse, error) {
+	nodeResponse, err := cs.StepClient.Recv()
+	if err != nil {
+		return nil, err
+	}
+	return BuildStepRespone(nodeResponse)
+}
+
+func (cs *NodeClientStream) Auth(version uint32, fromId uint64, toId uint64, channel string, signer identity.Signer) error {
+	if signer == nil {
+		return errors.New("signer is nil")
+	}
+
+	timestamp, err := ptypes.TimestampProto(time.Now().UTC())
+	if err != nil {
+		return errors.Wrap(err, "failed to read timestamp")
+	}
+
+	payload := &orderer.NodeAuthRequest{
+		Version:   version,
+		Timestamp: timestamp,
+		FromId:    fromId,
+		ToId:      toId,
+		Channel:   channel,
+	}
+
+	bindingFieldsHash := GetSessionBindingHash(payload)
+
+	tlsBinding, err := GetTLSSessionBinding(cs.StepClient.Context(), bindingFieldsHash)
+	if err != nil {
+		return errors.Wrap(err, "TLSBinding failed")
+	}
+	payload.SessionBinding = tlsBinding
+
+	asnSignFields, _ := asn1.Marshal(AuthRequestSignature{
+		Version:   int64(payload.Version),
+		Timestamp: payload.Timestamp.String(),
+		FromId:    strconv.FormatUint(payload.FromId, 10),
+		ToId:      strconv.FormatUint(payload.ToId, 10),
+		Channel:   payload.Channel,
+	})
+	sig, err := signer.Sign(asnSignFields)
+	if err != nil {
+		return errors.Wrap(err, "signing failed")
+	}
+
+	payload.Signature = sig
+	stepRequest := &orderer.ClusterNodeServiceStepRequest{
+		Payload: &orderer.ClusterNodeServiceStepRequest_NodeAuthrequest{
+			NodeAuthrequest: payload,
+		},
+	}
+
+	return cs.StepClient.Send(stepRequest)
+}
+
+func (cs *NodeClientStream) Context() context.Context {
+	return cs.StepClient.Context()
+}
+
+func BuildStepRequest(request *orderer.StepRequest) (*orderer.ClusterNodeServiceStepRequest, error) {
+	if request == nil {
+		return nil, errors.New("request is nil")
+	}
+	var stepRequest *orderer.ClusterNodeServiceStepRequest
+	if consReq := request.GetConsensusRequest(); consReq != nil {
+		stepRequest = &orderer.ClusterNodeServiceStepRequest{
+			Payload: &orderer.ClusterNodeServiceStepRequest_NodeConrequest{
+				NodeConrequest: &orderer.NodeConsensusRequest{
+					Payload:  consReq.Payload,
+					Metadata: consReq.Metadata,
+				},
+			},
+		}
+		return stepRequest, nil
+	} else if subReq := request.GetSubmitRequest(); subReq != nil {
+		stepRequest = &orderer.ClusterNodeServiceStepRequest{
+			Payload: &orderer.ClusterNodeServiceStepRequest_NodeTranrequest{
+				NodeTranrequest: &orderer.NodeTransactionOrderRequest{
+					Payload:           subReq.Payload,
+					LastValidationSeq: subReq.LastValidationSeq,
+				},
+			},
+		}
+		return stepRequest, nil
+	}
+	return nil, errors.New("service message type not valid")
+}
+
+func BuildStepRespone(stepResponse *orderer.ClusterNodeServiceStepResponse) (*orderer.StepResponse, error) {
```

Comment:
This function is not tested anywhere. I know we only use it in tests, so let's at least test it?